### PR TITLE
langchain[minor]: openai tools structured_output_chain

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/langchain/langchain/output_parsers/openai_tools.py [hunk 1]
```diff
@@ -22,6 +22,8 @@ class JsonOutputToolsParser(BaseGenerationOutputParser[Any]):
     """
     return_id: bool = False
     """Whether to return the tool call id."""
+    return_single: bool = False
```
Comment (eyurtsev) [inline libs/langchain/langchain/output_parsers/openai_tools.py:25]: This doesn't feel like an option that we should add to the parsers

File: libs/langchain/langchain/chains/structured_output/base.py [hunk 2]
```diff
@@ -106,15 +113,17 @@ class RecordDog(BaseModel):
...
 def create_structured_output_runnable(
     output_schema: Union[Dict[str, Any], Type[BaseModel]],
     llm: Runnable,
     prompt: BasePromptTemplate,
     *,
     output_parser: Optional[Union[BaseOutputParser, BaseGenerationOutputParser]] = None,
-    mode: Literal["openai-functions", "openai-json"] = "openai-functions",
-    enforce_single_function_usage: bool = True,
+    enforce_function_usage: bool = True,
+    return_single: bool = True,
```
Comment (eyurtsev) [inline libs/langchain/langchain/chains/structured_output/base.py:123]: Why is this option needed?

File: libs/langchain/langchain/output_parsers/openai_tools.py (hunk 1 above)
Comment (baskaryan) [reply libs/langchain/langchain/output_parsers/openai_tools.py:25]: why not

File: libs/langchain/langchain/chains/structured_output/base.py (hunk 2 above)
Comment (baskaryan) [reply libs/langchain/langchain/chains/structured_output/base.py:123]: openai tools returns a list otherwise

**Comment by ccurme** [comment]
Should we deprecate this? https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/chains/openai_tools/extraction.py

**Comment by baskaryan** [comment]
> Should we deprecate this? https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/chains/openai_tools/extraction.py

probably. think we should first land https://github.com/langchain-ai/langchain/pull/17302 and figure out what, if any, other extraction chains/utils we're adding, and then deprecate the various existing extraction chains

File: libs/langchain/langchain/chains/structured_output/base.py (hunk 2 above)
Comment (eyurtsev) [reply libs/langchain/langchain/chains/structured_output/base.py:123]: Should we maybe hide this from the user as an option? It feels like a weird option in the API

File: libs/langchain/langchain/output_parsers/openai_tools.py (hunk 1 above)
Comment (eyurtsev) [reply libs/langchain/langchain/output_parsers/openai_tools.py:25]: Do you mind if we re-name this to: "first_tool_only" so it's clear how to interpret "single"

File: libs/langchain/langchain/chains/structured_output/base.py (hunk 2 above)
Comment (eyurtsev) [reply libs/langchain/langchain/chains/structured_output/base.py:123]: Can keep this as `return_single`

File: libs/langchain/langchain/output_parsers/openai_tools.py [hunk 3]
```diff
@@ -22,8 +22,16 @@ class JsonOutputToolsParser(BaseGenerationOutputParser[Any]):
     """
     return_id: bool = False
     """Whether to return the tool call id."""
-    return_single: bool = False
-    """Whether to return only the first tool call."""
+    first_tool_only: bool = False
```
Comment (baskaryan) [inline libs/langchain/langchain/output_parsers/openai_tools.py:25]: breaking

Comment (baskaryan) [reply libs/langchain/langchain/output_parsers/openai_tools.py:25]: should probs alias

File: libs/langchain/langchain/chains/structured_output/base.py [hunk 4]
```diff
@@ -299,33 +306,46 @@ class Dog(BaseModel):
                 chain = create_structured_output_runnable(Dog, llm, prompt, mode="openai-json")
                 chain.invoke({"input": "Harry was a chubby brown beagle who loved chicken"})
     """  # noqa: E501
+    # for backwards compatibility
```
Comment (baskaryan) [inline libs/langchain/langchain/chains/structured_output/base.py:320]: i don't think we want to support this outside of openai-functions?

File: libs/langchain/langchain/output_parsers/openai_tools.py (hunk 3 above)
Comment (eyurtsev) [reply libs/langchain/langchain/output_parsers/openai_tools.py:25]: Why is this breaking? is this used somewhere already?

File: libs/langchain/langchain/chains/structured_output/base.py (hunk 4 above)
Comment (eyurtsev) [reply libs/langchain/langchain/chains/structured_output/base.py:320]: I can push inside, but I usually handle this at top scope, even better would be to fix this via a function level decorator to add a parameter renamer

File: libs/langchain/langchain/output_parsers/openai_tools.py (hunk 3 above)
Comment (baskaryan) [reply libs/langchain/langchain/output_parsers/openai_tools.py:25]: L76

File: libs/langchain/langchain/output_parsers/openai_tools.py [hunk 5]
```diff
@@ -73,21 +85,36 @@ class JsonOutputKeyToolsParser(JsonOutputToolsParser):
 
     key_name: str
     """The type of tools to return."""
-    return_single: bool = False
-    """Whether to return only the first tool call."""
-
     def __init__(self, key_name: str, **kwargs: Any) -> None:
         """Allow init with positional args."""
+        # Backwards compatibility for old argument name.
+        if "return_single" in kwargs:
```
Comment (eyurtsev) [inline libs/langchain/langchain/output_parsers/openai_tools.py:91]: Adding backwards compatibility

File: libs/langchain/langchain/chains/structured_output/base.py [hunk 6]
```diff
@@ -130,19 +140,107 @@ def create_structured_output_runnable(
...
+                    fav_food: Optional[str] = Field(None, description="The dog's favorite food")
 
-    OpenAI functions example:
+                llm = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0)
+                prompt = ChatPromptTemplate.from_messages(
+                    [
+                        ("system", "You are an extraction algorithm. Please extract every possible instance"), 
+                        ('human', '{input}')
+                    ]
+                )
```
Comment (baskaryan) [inline libs/langchain/langchain/chains/structured_output/base.py:185]: would be nice to have an example without prompt

**Pull request merged**